Adaptive VQE

mrossinek opened the pull request:
<!--
⚠️ If you do not respect this template, your pull request will be closed.
⚠️ Your pull request title should be short detailed and understandable for all.
⚠️ Also, please add it in the CHANGELOG file under Unreleased section.
⚠️ If your pull request fixes an open issue, please link to the issue.

✅ I have added the tests to cover my changes.
✅ I have updated the documentation accordingly.
✅ I have read the CONTRIBUTING document.
-->

### Summary
This PR implements an adaptive version of the VQE algorihtm as proposed here: https://arxiv.org/abs/1812.11173
I would like to note in the very beginning, that this contribution is the result of a team effort at the Qiskit Camp Europe 2019: qiskit-community/qiskit-camp-europe-19#39

A few results which prove that the implementation works as expected can be found [at this tag](https://github.com/mrossinek/qiskit-aqua/tree/qiskit-camp-europe-19). (see the `.dat` files)

### Details and comments
Some details regarding the implementation:
- The new algorithm is implemented in the `vqe_adapt` module. The class `VQEAdapt` derives from `VQAlgorithm` and implements the basic iterative procedure.
- A new variational form `UCCSDAdapt` has been added which derives from `UCCSD` and provides two additional methods to allow the iterative addition and removal of hopping operators.

Required changes since after the Qiskit Camp ended:
- [x] clean up the repository of unwanted files (temporary results, testing scripts, etc.)
- [x] rebase the work on `master` rather than `stable`
- [x] simplify the `UCC` implementation (now `UCCSDAdapt`)
- [x] clean up the code in general (current work in progress)
- [x] use logger instead of simply printing debug output
- [x] implement missing getter functions (optimal cost, circuit, vector, params) [open for discussion]
- [x] write unittests
- ~~[ ] provide a notebook to showcase the usage (will probably have to be added to the advanced community tutorials) (will probably be based on [this one](https://github.com/mrossinek/qiskit-aqua/blob/qiskit-camp-europe-19/UCCSD_VQE_reference.ipynb))~~
**Edit:** I would consider this to go into the community tutorials.

We had some further ideas on how this current implementation can be generalized even further to work with other variational forms than UCCSD but for now I would propose integrating this basic implementation and extending it when needed.

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/not_signed)](https://cla-assistant.io/Qiskit/qiskit-aqua?pullRequest=685) <br/>Thank you for your submission, we really appreciate it. Like many open source projects, we ask that you all sign our [Contributor License Agreement](https://cla-assistant.io/Qiskit/qiskit-aqua?pullRequest=685) before we can accept your contribution.<br/>**1** out of **2** committers have signed the CLA.<br/><br/>:white_check_mark: mrossinek<br/>:x: arsenyko<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla-assistant.io/check/Qiskit/qiskit-aqua?pullRequest=685) it.</sub>

mrossinek commented:
Okay so I finally got around to fixing the spelling errors and I also made both new classes pluggable types. Although I am not sure if that is done properly. The Travis build still failed due to a different test case which is not part of this PR.

mrossinek commented:
Per my last edit of this PRs description I believe that the jupyter notebook, which is meant to be more educational and as a showcase, should be added to the [community tutorials](https://github.com/Qiskit/qiskit-community-tutorials) rather than to the code of qiskit aqua itself.

mrossinek commented on qiskit/chemistry/aqua_extensions/algorithms/adaptive/vqe_adapt/vqe_adapt.py:
```diff
@@ -0,0 +1,269 @@
+            energy_results = vqe._energy_evaluation(np.asarray(parameter_sets))
+            # compute real and imaginary gradients
+            gradient = (energy_results[0] - energy_results[1]) / (2*delta)
+            gradient_i = (energy_results[2] - energy_results[3]) / (2*delta)
```
```suggestion
```

mrossinek commented on qiskit/chemistry/aqua_extensions/algorithms/adaptive/vqe_adapt/vqe_adapt.py:
```diff
@@ -0,0 +1,269 @@
+            gradient = (energy_results[0] - energy_results[1]) / (2*delta)
+            gradient_i = (energy_results[2] - energy_results[3]) / (2*delta)
+            # for now: simply use maximum of either
+            res.append((max(np.abs(gradient), np.abs(gradient_i)), exc))
```
```suggestion
            res.append((np.abs(gradient), exc))
```

mrossinek commented on qiskit/chemistry/aqua_extensions/algorithms/adaptive/vqe_adapt/vqe_adapt.py:
```diff
@@ -0,0 +1,269 @@
+            # compute real and imaginary gradients
+            gradient = (energy_results[0] - energy_results[1]) / (2*delta)
+            gradient_i = (energy_results[2] - energy_results[3]) / (2*delta)
+            # for now: simply use maximum of either
```
```suggestion
```

mrossinek commented on qiskit/chemistry/aqua_extensions/algorithms/adaptive/vqe_adapt/vqe_adapt.py:
```diff
@@ -0,0 +1,269 @@
+            parameter_sets = theta + [-delta] + theta + [delta] \
+                + theta + [-delta*1j] + theta + [delta*1j]
+            energy_results = vqe._energy_evaluation(np.asarray(parameter_sets))
+            # compute real and imaginary gradients
```
```suggestion
            # compute gradient
```

mrossinek commented on qiskit/chemistry/aqua_extensions/algorithms/adaptive/vqe_adapt/vqe_adapt.py:
```diff
@@ -0,0 +1,269 @@
+            vqe._quantum_instance = self._quantum_instance
+            vqe._use_simulator_operator_mode = self._use_simulator_operator_mode
+            # evaluate energies
+            parameter_sets = theta + [-delta] + theta + [delta] \
```
```suggestion
            parameter_sets = theta + [-delta] + theta + [delta]
```

woodsp-ibm commented:
Thanks for the contribution and the algorithm. Apologies for delay in getting back to you. I have a few comments.

1. We are thinking that the UCCSDAdapt changes should be made part of UCCSD. UCCSDAdapt will not work outside of VQEAdapt, for instance if just used with VQE. Now while we could make UCCSDAdapt default to behave just like UCCSD, unless its hopping ops set was wanted to be managed, e.g by VQEAdapt, it seems we may as well extend UCCSD to allow it to be managed. This simplifies things from declarative too. Now we would just have UCCSD, that can used in VQE would work just the same as it does today, but if used in VQEAdapt the adaptive behavior can take place.

As such we would need one more method, in addition to push and pop which would more or less be whats in the constructor. Maybe its just called reset_hopping_operators which empties the list and then allows push and pop (maybe push and pop is better than append and pop). So when VQEAdapt checks that its passed a UCCSD instance (its UCCSDAdapt now) it would first call the reset method to ensure things start as expected so the hopping operators can then be managed by it.

2. In pop, should it be called and the hopping operators are empty, it still decrements num_parameters. Maybe it would be better, for both push and pop that num_parameters is just set to the length of the hopping operators in those methods. This will ensure it stays consistent. Now in UCCSD we have a depth parameter so the 'normal' num parameters is the num hopping operators multipled by depth. And I think here those methods should do the same. Now this adaptive mode may only make real sense for depth=1, in which case I think the depth passed into UCCSD should be checked, maybe in the reset_hopping_operators, though now I am liking the name less, and if not 1 then a warning logged and the depth overridden to 1. This will prevent any issues around incorrect config. With this extra logic maybe the proposed method reset_hopping_operators should be manage_hopping_operators which will do this depth test and reset the list etc.

3. In VQEAdapt there is a parameter 'excitation_pool' in the constructor. I would propose to default this to None and if None then VQEAdapt gets the excitation_pool from the UCCSD that is passed. I am assuming the parameter exists so that the pool can be customized beyond what is possible say via the active space settings in UCCSD and a user can then pass that in for use. This way by default the user is not required to get the pool from UCCSD and pass it across, which is how the code is today. This will also allow things to be run declaratively as well as being simpler for default case.

4. For the CONFIGURATION this should look pretty much like VQE but without the max_evals_grouped I think, unless you want to be able to allow users to set that whereupon you would pass it to the VQE you construct internally. The depends in the CONFIG should be an optimizer and var form - the same as is required in VQEAdapt constructor, i.e what a user is required to pass in. (VQE is an internal hard-coded dependence and not something that can be passed in so should not be in there). I think threshold and delta should be in the CONFIGURATION so that all the params from VQEAdapt constructor are covered.

Also VQEAdapt should have an init_params classmethod that would look much like the classmethod by that name in VQE. I think the only difference would be that max_evals_group setting gets removed (according to above discussion) and threshold and delta are added.

I can provide more help here if needed to do this CONFIG and init_params, if the above is unclear.

5. VQE aux_ops. This is a parameter containing an array of what we call auxiliary operators. The contract with an algorithm that can process these is that they should be evaluated against the solution ground state. We use these to pass through operators such as dipole, spin and particle count. VQE has no idea what they represent it just evaluates them and passes the result back for the caller to interpret. You can see we often use these to check spin, particles etc. It would be nice to support these in VQEAdapt too, so that VQE can evaluate them once the adaption is done. As such I was thinking that if aux_ops can be passed to VQEAdapt and was not None and not empty that once VQEAdapt terminated at a solution it could run VQE one more time but this time including the aux_ops to have those evaluated too as part of the final result.

mrossinek commented:
@woodsp-ibm Hi Steve! Thanks for taking the time to look at my PR. I am glad about your feedback and have inserted some comments and links to the relevant commits below each of your sections below.

> 1. We are thinking that the UCCSDAdapt changes should be made part of UCCSD. UCCSDAdapt will not work outside of VQEAdapt, for instance if just used with VQE. Now while we could make UCCSDAdapt default to behave just like UCCSD, unless its hopping ops set was wanted to be managed, e.g by VQEAdapt, it seems we may as well extend UCCSD to allow it to be managed. This simplifies things from declarative too. Now we would just have UCCSD, that can used in VQE would work just the same as it does today, but if used in VQEAdapt the adaptive behavior can take place.

I merged the `UCCSDAdapt` class into `UCCSD` in commit 5f2a622.

> 2. In pop, should it be called and the hopping operators are empty, it still decrements num_parameters. Maybe it would be better, for both push and pop that num_parameters is just set to the length of the hopping operators in those methods. This will ensure it stays consistent. Now in UCCSD we have a depth parameter so the 'normal' num parameters is the num hopping operators multipled by depth. And I think here those methods should do the same. Now this adaptive mode may only make real sense for depth=1, in which case I think the depth passed into UCCSD should be checked, maybe in the reset_hopping_operators, though now I am liking the name less, and if not 1 then a warning logged and the depth overridden to 1. This will prevent any issues around incorrect config. With this extra logic maybe the proposed method reset_hopping_operators should be manage_hopping_operators which will do this depth test and reset the list etc.

I addressed this change in fdc6e68.

> 3. In VQEAdapt there is a parameter 'excitation_pool' in the constructor. I would propose to default this to None and if None then VQEAdapt gets the excitation_pool from the UCCSD that is passed. I am assuming the parameter exists so that the pool can be customized beyond what is possible say via the active space settings in UCCSD and a user can then pass that in for use. This way by default the user is not required to get the pool from UCCSD and pass it across, which is how the code is today. This will also allow things to be run declaratively as well as being simpler for default case.

See 7e59afb.

> 4. For the CONFIGURATION this should look pretty much like VQE but without the max_evals_grouped I think, unless you want to be able to allow users to set that whereupon you would pass it to the VQE you construct internally. The depends in the CONFIG should be an optimizer and var form - the same as is required in VQEAdapt constructor, i.e what a user is required to pass in. (VQE is an internal hard-coded dependence and not something that can be passed in so should not be in there). I think threshold and delta should be in the CONFIGURATION so that all the params from VQEAdapt constructor are covered.
> Also VQEAdapt should have an init_params classmethod that would look much like the classmethod by that name in VQE. I think the only difference would be that max_evals_group setting gets removed (according to above discussion) and threshold and delta are added.
> I can provide more help here if needed to do this CONFIG and init_params, if the above is unclear.

I tried my best to implement this in 7a15c4e. I think I got it all but the code could benefit from another check by you.

> 5. VQE aux_ops. This is a parameter containing an array of what we call auxiliary operators. The contract with an algorithm that can process these is that they should be evaluated against the solution ground state. We use these to pass through operators such as dipole, spin and particle count. VQE has no idea what they represent it just evaluates them and passes the result back for the caller to interpret. You can see we often use these to check spin, particles etc. It would be nice to support these in VQEAdapt too, so that VQE can evaluate them once the adaption is done. As such I was thinking that if aux_ops can be passed to VQEAdapt and was not None and not empty that once VQEAdapt terminated at a solution it could run VQE one more time but this time including the aux_ops to have those evaluated too as part of the final result.

In 6fe643b I added the `aux_operators` and `max_eval_grouped` optional arguments to `VQEAdapt` in order to be more inline with `VQE`. Please have a look at [line 292-296](https://github.com/mrossinek/qiskit-aqua/blob/adapt-vqe-master/qiskit/chemistry/aqua_extensions/algorithms/adaptive/vqe_adapt/vqe_adapt.py#L292) if this is what you had in mind.

mrossinek commented on qiskit/chemistry/aqua_extensions/algorithms/adaptive/vqe_adapt/vqe_adapt.py:
```diff
@@ -0,0 +1,269 @@
+            vqe._use_simulator_operator_mode = self._use_simulator_operator_mode
+            # evaluate energies
+            parameter_sets = theta + [-delta] + theta + [delta] \
+                + theta + [-delta*1j] + theta + [delta*1j]
```
For completeness: all these changes were resolved in 944317f.

woodsp-ibm commented:
Thanks for the changes. A couple of final comments then that hopefully wraps things up.

The CONFIGURATION in VQEAdapt looks fine except that we cannot reasonably handle excitation_pool via the declarative json path as the array contents would be too complex to do that way - but since excitation_pool defaults to None in VQEAdapt constructor it will use what is in UCCSD which will be fine for declarative path.  If one wanted to do something more advanced then it would have to be via direct programming in Python. So these lines should vanish from the CONFIGURATION 
```
                'excitation_pool': {
                    'type': ['array', 'null'],
                    "items": {
                        "type": "WeightedPauliOperator"
                    },
                    'default': None
                },
```

And one last more minor thing. Since the _manage/_push/_pop_hopping_operator methods are accessed from VQEAdapt outside of UCCSD it would be better that the _ private indicator be removed from these so its just manage/push/pop_.... Bear in mind being public out lint/style check will now require a full docstring describing parameters, returns etc.

mrossinek commented:
Thanks for the new feedback! Should all be implemented now.
Since none of the three functions have a return value the docstring simplifies, correct? At least locally the `make lint` and `make style` did not produce any errors for me.

woodsp-ibm commented:
Right the docstring change for the functions should be simple. I guess I was talking more generically that public functions need all that stuff documented - as long as make lint/style passes locally, as yours did, then it should pass the build fine.

Thanks again for the contribution and making the changes

woodsp-ibm commented:
Oh, please note that once this PR is merged we will make a small change via PR #700 in which we have refactored some logic and removed the aqua_extensions folder during the process, placing algorithms and components directly under the domain folder i.e. chemistry, finance etc. Just in case you wonder what happened! We held that PR pending this being merged and we'll take care of the folder change in that other PR.

mrossinek commented:
Okay thank you for the heads up! I will have to make sure to update the pending community tutorial once that PR is merged then.

mrossinek commented:
Oh and one more thing: I did not add anything about this PR to the Changelog. I totally forgot about that. Could you do that in a single commit or should we go through another PR for that to happen?

woodsp-ibm commented:
My bad too! I did not notice. We'll take care of that and add some text for this in the PR I referenced above to save doing another PR just for the changelog mod.
